tools/migrate_encrypted_value_to_db.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+
+# Copyright 2020 The StackStorm Authors.
+# Copyright 2019 Extreme Networks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+from oslo_config import cfg
+
+from mongoengine.queryset import Q
+import binascii
+
+from hashlib import sha1
+import six
+
+from st2common import config
+from st2common.script_setup import setup as db_setup
+from st2common.script_setup import teardown as db_teardown
+from st2common.persistence.keyvalue import KeyValuePair
+from st2common.models.db.keyvalue import KeyValuePairDB
+
+from st2common.util.crypto import AESKey
+from st2common.util.crypto import symmetric_encrypt
+from st2common.util.crypto import read_crypto_key
+from st2common.util.crypto import Base64WSDecode
+from st2common.util.crypto import pkcs5_unpad
+
+
+from cryptography.hazmat.primitives.ciphers import Cipher
+from cryptography.hazmat.primitives.ciphers import algorithms
+from cryptography.hazmat.primitives.ciphers import modes
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import hmac
+from cryptography.hazmat.backends import default_backend
+
+# Keyczar related constants
+KEYCZAR_HEADER_SIZE = 5
+KEYCZAR_AES_BLOCK_SIZE = 16
+KEYCZAR_HLEN = sha1().digest_size
+
+
+class EncryptedValueMigration(object):
+    def _get_keyvalue_with_parameters(self):
+        """
+        All KeyValueDB that has a secret parameter.
+        """
+        return KeyValuePairDB.objects(Q(secret=True))
+
+    def migrate_encrypted_value(self):
+        """
+        Will migrate the new encrypted value to KeyValuePair.
+        """
+        keyvalue_dbs = self._get_keyvalue_with_parameters()
+        crypto_key_path = cfg.CONF.keyvalue.encryption_key_path
+        crypto_key = read_crypto_key(crypto_key_path)
+        hmac_key_bytes = Base64WSDecode(crypto_key.hmac_key_string)
+        aes_key_bytes = Base64WSDecode(crypto_key.aes_key_string)
+        for keyvalue_db in keyvalue_dbs:
+            if keyvalue_db.secret:
+                plaintext = cryptography_symmetric_decrypt(
+                    hmac_key_bytes, aes_key_bytes, keyvalue_db.value
+                )
+                crypto_key = AESKey.generate()
+                new_encrypted_value = symmetric_encrypt(crypto_key, plaintext)
```

Comment:
Why are we generating a new crypto_key for every keyvalue entry?
I would imagine that we should generate the new key once, and save it to some path, and then use that key to reencrypt all the values.